[LaTeX] Text formatting

ngc92 opened the pull request:
The PR reworks the handling of text formatting commands in LaTeX. In particular, in cases where there are commands within formatted text, e.g. `\textbf{\TeX in bold}`, we no longer format the command names. 

One challenge is that there are several font properties that can be changed independently: Font family (roman, sans serif, typewriter), font weight (bold, medium, (possibly) light), and font shape (upright, italic, slanted, small-caps). Most of these do not have any corresponding `markup` scopes and highlighting rules (at least with default color schemes). At this point, only a subset of nesting of these properties is actually handled: Bold/italic combination, and nested \emph emphasize, as well as ensuring that all the commands that do not change one of these properties still typeset their arguments according to the outer command.

No provision has been made for any form of triple-nesting, i.e. `\textbf{\textit{\textup{a}}}` would not result in bold `a` right now. Not sure if supporting this would be worth it.

Finally, there is also an \underline command that could be combined with bold/italic. So far, I've not included this, because in LaTeX underline comes with some caveats, so that package commands like `\uline` or `\ul` are often recommended.

A more general question, is there a way to generate variations of the same context without excessive copy-and-paste? I've seen that there is a json-macro plugin, but I'm not sure if this is officially supported.

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +702,382 @@ contexts:
+      scope: punctuation.definition.group.brace.begin.tex
+      push: text-bold-brace-group-body
+    - include: main
+    - match: .
```
This pattern causes ST to create a token for each single character.

See how each char of "bold" and "still bold" is highlighted differently.

![grafik](https://user-images.githubusercontent.com/16542113/213877196-6cdd50a8-17be-4953-a3a3-b7084e2c8e1e.png)

A possible solution/improvement is to use a pattern which matches all chars, which the other patterns are not expected to start with.

The following pattern:

```suggestion
    - match: '[^\\{}]+'
```

produces more sane results:

![grafik](https://user-images.githubusercontent.com/16542113/213877340-fb1da26b-a248-4eb5-943a-b0c175c7714f.png)

_Note: Effects the other two occurances in `italic` and `bold-italic` as well._

ngc92 replied:
its a bit more complicated than that, e.g. you suggestion would break `\textbf{bold $normal math$}` (which is currently missing from the tests)

deathaxe replied:
I'd vote to add tests for some of the more exotic cases as well in general. As those are often root cause of breaking highlighting _(lessons learned)_.

I already was concerned about such cases.

If in doubt we can go with your solution - it's not wrong - just not optimal from performance/memory concumption point of view.

michaelblyons replied:
`\w+|.`?

deathaxe replied:
If all special chars were known those could be added to the char class `[^\\{}%$]+`. Otherwise we could refine your suggestion to something like `.[\s\w]*` maybe.

deathaxe replied:
If it was acceptable to also scope keywords `markup.bold`, ... a `meta_content_scope` may be used as well to scope everything within braces accordingly.

ngc92 replied:
I think I prefer the more refined match over making everything `markup.bold`. Explicitly opting out only for a few characters might work, because
a) even if they have special meaning in latex, it really hurts only for those for which we do specific processing
b) any sort of fancy latex code (directly) inside text markup is very unlikely. E.g., the `#`, `_`, `^`, `&` characters would be problematic, but I think they aren't actually possible inside well-formed latex, so even if we do get these wrong, nothing too bad would be happening.

deathaxe commented:
The rest looks good to me. Nice modular context structure.

ngc92 commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +702,382 @@ contexts:
-        - meta_scope: meta.function.textbf.latex
-        - meta_content_scope: markup.bold.textbf.latex
+        - meta_scope: meta.function.underline.latex
+        - meta_content_scope: markup.underline.underline.latex
```
possibly. Maybe I wanted to differentiate it from `markup.underline.link.latex`

deathaxe replied:
`markup.underline.underline.latex` has been used in LaTeX before.

If we want to distinguish links and normal underlines via 3rd-level scope, I'd suggest `markup.underline.other.latex` as `other` is a rather general pattern used in various situations for common stuff.

ngc92 commented:
> Not sure what that `./.meta/blacklist.txt` is for. Seems unrelated to this PR anyways.

Sorry, should not have been added. That was just me taking notes as part of my script for #3607

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -15,6 +15,9 @@ first_line_match: |-
   )
 
+variables:
+  simpletext: '.[^\\{}%$~&#_^]+'
```
```suggestion
  simpletext: '.[^\\{}%$~&#_^]*'
```

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +705,382 @@ contexts:
+      scope: punctuation.definition.group.brace.begin.tex
+      push: text-bold-brace-group-body
+    - include: main
+    - match: .{{simpletext}}
```
```suggestion
    - match: '{{simpletext}}'
```

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +705,382 @@ contexts:
+        - text-normal-body
+        - brace-group-begin
+    - include: main
+    - match: .{{simpletext}}
```
```suggestion
    - match: '{{simpletext}}'
```

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +705,382 @@ contexts:
+  text-it-and-bf-body:
+    - include: brace-group-end
+    - include: main
+    - match: .{{simpletext}}
```
```suggestion
    - match: '{{simpletext}}'
```

deathaxe commented on LaTeX/LaTeX.sublime-syntax:
```diff
@@ -702,75 +705,382 @@ contexts:
+  texttt-body:
+    - include: brace-group-end
+    - include: main
+    - match: .{{simpletext}}
```
```suggestion
    - match: '{{simpletext}}'
```

deathaxe reviewed (commented):
1. `simpletext` should match on single chars as well, hence changing `+` to `*`.
2. as `simpletext` variable already starts with `.` it is redundant when using it, hence changing patterns.

ngc92 commented:
I've fixed the matching, though in the other way around. I think that makes the name "simpletext" more fitting for the variable, as now it only matches the "simple" part, instead of matching an arbitrary character followed by simple things.

I've also added a test case that would have broken with the old pattern.

deathaxe commented:
Reorganized contexts a bit and changed `underline.underline` to `underline.other`. Would you please double check my changes to make sure I haven't introduced some regressions?

If you are fine with it, I guess we can merge it.

ngc92 commented:
One question is if we want to treat underline like the other text commands. With the current setup, it provides nested underline/bold etc nicely, but it also underlines commands, comments etc. within its scope. This is especially noticeable with incomplete text, e.g. if you have `\underline{typing some text in the middle of the document`, and don' t have a closing brace yet, then the entire rest of the document ends up being underline.

deathaxe commented:
I was thinking about the same during review. But wouldn't applying the strategy of the other contexts mean to add a lot of more contexts to make sure all nested formattings are scoped as "underlined" properly? I don't think it's worth the complexity. Incomplete statements may always break highlighting in some way.
